### PR TITLE
fix(NODE-6372): correctly GSS wrap non-user data

### PR DESCRIPTION
**Pull request by dsullivan**
### Description

Fix implementation of `wrap` function to work for use cases other than authentication.

#### What is changing?

Just an extra check for an empty string: `if (user)` becomes `if (user && *user)`.

Without this, the `wrap` function was discarding the challenge/payload value and wrapping only the user option value, even if no such value was provided.

##### Is there new documentation needed for these changes?

No, this makes the `wrap` method work as per the existing documentation.

#### What is the motivation for this change?

Without this change, the `wrap` function works only for authentication. With this change, any payload can be GSS-wrapped.

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the [correct format](https://www.conventionalcommits.org/en/v1.0.0/): `type(NODE-xxxx)[!]: description`
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
- [x] Changes are covered by tests
- [x] New TODOs have a related JIRA ticket

### CODE REVIEW

**Comment by dsullivan** [comment]
NOTE: I wasn't able to run the full test suite locally, given my Kerberos server configuration. I did run equivalent versions of the new tests in my application.

File: test/kerberos_tests.js [hunk 1]
```diff
@@ -168,5 +168,24 @@ describe('Kerberos', function () {
         });
       });
     });
+
+    describe('options.user', function () {
+      context('valid values for `user`', function () {
+        test('no options provided', async function () {
+          const rs = await client.wrap('x'.repeat(100));
+          expect(rs).length.to.be.greaterThan(100);
```
Comment (dsullivan) [inline test/kerberos_tests.js:176]: We noticed the bug because the wrapped value was actually shorter than the original challenge text. This is just a simple assertion to demonstrate that the result is plausible.

When we _are_ providing a user, the challenge text is still discarded and therefore the result is shorter.

**Review by nbbeeken — APPROVED** [review]
Thanks for the contribution with tests!

File: src/unix/kerberos_gss.cc [hunk 2]
```diff
@@ -468,7 +468,7 @@ gss_result authenticate_gss_client_wrap(gss_client_state* state,
         input_token.length = len;
     }
 
-    if (user) {
+    if (user && *user) {
```
Comment (addaleax) [inline src/unix/kerberos_gss.cc:471]: Should the check in L519 also be adjusted in the same way? The memory management code in this file is honestly a bit hard to follow, but it seems that way to me

Comment (dsullivan) [reply src/unix/kerberos_gss.cc:471]: I'm also having a hard time following.

My suspicion is that, since `user` is never null, that release is never being executed. Or at least has not been since the `ToStringWithNonStringAsEmpty` behavior was introduced.

Then again, I don't understand why we'd only need to release the buffer if there is no user. Maybe the right fix is to remove that condition altogether and just call the release if there is any data in the buffer?

Comment (addaleax) [reply src/unix/kerberos_gss.cc:471]: > Then again, I don't understand why we'd only need to release the buffer if there is no user

I assume that it corresponds to this condition here, hence my suggestion above – if this branch is taken, then `input_token` contains a stack-allocated reference to `buf`, rather than a dynamically allocated buffer

**Comment by nbbeeken** [comment]
Closing for inactivity, please feel free to re-open and continue this if there's interest!
